precompiles/assets-erc20/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+			output: vec![],
+			logs: LogsBuilder::new(context.address)
+				.log3(
+					SELECTOR_LOG_APPROVAL,
+					context.caller,
+					spender,
+					EvmDataWriter::new().write(amount).build(),
+				)
+				.build(),
+		})
+	}
+
+	fn transfer(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_log_costs_manual(3, 32)?;
+
+		// Parse input.
+		input.expect_arguments(2)?;
+
+		let to: H160 = input.read::<Address>()?.into();
+		let amount: U256 = input.read()?;
+
+		// Build call with origin.
+		{
+			let execution_address = Runtime::AddressMapping::into_account_id(context.address);
+			let asset_id: AssetIdOf<Runtime, Instance> =
+				Runtime::Precompiles::account_to_asset_id(execution_address)
+					.ok_or(error("non-assetId address"))?;
+
+			let origin = Runtime::AddressMapping::into_account_id(context.caller);
+			let to = Runtime::AddressMapping::into_account_id(to);
+			let amount = Self::u256_to_amount(amount)?;
+
+			// Dispatch call (if enough gas).
+			let used_gas = RuntimeHelper::<Runtime>::try_dispatch(
+				Some(origin).into(),
+				pallet_assets::Call::<Runtime, Instance>::transfer(
+					asset_id,
+					Runtime::Lookup::unlookup(to),
+					amount,
+				),
+				gasometer.remaining_gas()?,
+			)?;
+			gasometer.record_cost(used_gas)?;
+		}
+
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: vec![],
+			logs: LogsBuilder::new(context.address)
+				.log3(
+					SELECTOR_LOG_TRANSFER,
+					context.caller,
+					to,
+					EvmDataWriter::new().write(amount).build(),
+				)
+				.build(),
+		})
+	}
+
+	fn transfer_from(
+		mut input: EvmDataReader,
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> EvmResult<PrecompileOutput> {
+		let mut gasometer = Gasometer::new(target_gas);
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		gasometer.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
+		gasometer.record_log_costs_manual(3, 32)?;
+
+		// Parse input.
+		input.expect_arguments(3)?;
+		let from: H160 = input.read::<Address>()?.into();
+		let to: H160 = input.read::<Address>()?.into();
+		let amount: U256 = input.read()?;
+
+		{
+			let execution_address = Runtime::AddressMapping::into_account_id(context.address);
+			let asset_id: AssetIdOf<Runtime, Instance> =
+				Runtime::Precompiles::account_to_asset_id(execution_address)
+					.ok_or(error("non-assetId address"))?;
+			let caller: Runtime::AccountId =
+				Runtime::AddressMapping::into_account_id(context.caller);
+			let from: Runtime::AccountId = Runtime::AddressMapping::into_account_id(from.clone());
+			let to: Runtime::AccountId = Runtime::AddressMapping::into_account_id(to);
+			let amount = Self::u256_to_amount(amount)?;
+
+			// If caller is "from", it can spend as much as it wants.
+			if caller != from {
```

Comment:
```suggestion
			let used_gas = if caller != from {
```
Then you can update the gasometer after the `if`.